### ISSUE TITLE
Ending AudioEngine in Win32 causes a "std::system_exception"

### ISSUE THREAD

**Issue by Gajoo** [cocos2d/cocos2d-x#13523]
I'm using cocos2d 3.7.1. It's recommended in documentation to "end" AudioEngine before the app exits, so Here is what I did:

```
AppDelegate::~AppDelegate() 
{
    cocos2d::experimental::AudioEngine::end();
}
```

But that line results in an exception being thrown on line 95 of "AudioEngine-win32.cpp". On that line AudioEngineThreadPool is trying to join all threads running, to safely exit the application. It's "std::system_exception" error, caused by threads not being joinable.

```
void destroy()
{
    _running = false;
    _sleepCondition.notify_all();

    for (int index = 0; index < _numThread; ++index) {
        _threads[index].join();
    }
}           
```

After digging more into the code, I've found this, inside AudioEngineThreadPool constructor (line 65 of same file):

```
_threads.reserve(_numThread);
_tasks.reserve(_numThread);

for (int index = 0; index < _numThread; ++index) {
    _tasks.push_back(nullptr);
    _threads.push_back( std::thread( std::bind(&AudioEngineThreadPool::threadFunc,this,index) ) );
    _threads[index].detach();
}
```

On line 65, all threads are detached, meaning they are no longer joinable. But in destroy function, it tries to join all those threads, even though they are detached aka. non-joinable; Which according to C++ specification should throw system_error exception [[reference]](http://www.cplusplus.com/reference/thread/thread/join/)

Potential fixes:
1. Remove detach from line 65
2. check if thread is joinable, before joining it on line 95

**Comment by Gajoo** [comment]
Looking a little bit more into the code, it seems on line 82 also, there is a possibility that a new thread is created, and detached. So if you are going with suggestion one, you should also remove detach from that line too.

**Comment by Gajoo** [comment]
After testing both methods, it seems adding a check if thread are joinable on line 95, will fix this problem but will cause another exception. So the only working solution is removing detach from line 65 and 82

**Comment by WenhaiLin** [comment]
Fixed in v3.8

**Comment by ghost** [comment]
I am using cocos2d-x v3.12 and I have the same exact issuein Win32, I think it is not fixed !

The program crashes when doing:

```
AppDelegate::~AppDelegate() 
{
    cocos2d::experimental::AudioEngine::end();
}
```

The program generally crashes in AudioEngine.cpp in threadFunc(), on _taskCondition.wait(lk); at line 127, with an access violation error

```
void threadFunc()
    {
        while (true) {
            std::function<void()> task = nullptr;
            {
                std::unique_lock<std::mutex> lk(_queueMutex);
                if (_stop)
                {
                    break;
                }
                if (!_taskQueue.empty())
                {
                    task = std::move(_taskQueue.front());
                    _taskQueue.pop();
                }
                else
                {
                    _taskCondition.wait(lk);
                    continue;
                }
            }

            task();
        }
    }
```

It looks like something was destroyed while the thread was still running.
Probably when destroying the AudioEngineThreadPool, it destroys variable without waiting for the workers to be finished ?

**Comment by dumganhar** [comment]
@graphiboc , thanks for your feedback. I'll check it.

**Comment by dumganhar** [comment]
Reopened this issue.

**Comment by k-akiyama1991** [comment]
I'm using ver 3.13.1 and I get the same problem in win32.

I agree with **@graphidoc**, the program clashes because these threads attempt to access a member valiable which was already destroied by `AudioEngineThreadPool`'s destructor.

I'm newbie of C++ thread programming, so someone please tell me, why should we detach these threads?
If we don't do that, the destructor can wait these threads after turn on the `_stop` flag until they complete thier job.

**Comment by dumganhar** [comment]
I think this issue was fixed in v3.14.1. 
I closed this issue.
If you find that it still exists, feel free to re-open it.
